Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -562,102 +536,133 @@ type safety to make its capture type be a homogeneous tuple instead of an array,
+### `Regex<Captures>` instead of `Regex<Match>`
 
-However, a `Never` value never exists. Functions with return type `Never` will
-never return. As a result, calling an API like `captures` on a regex with no
-captures would cause the program to abort or hang.
+In the initial version of this pitch, `Regex` was _only_ generic over its
+captures and `firstMatch(of:)` was responsible for flattening together the
+match and captures into a tuple.
 
 ```swift
-let identifier = /[_a-zA-Z]+[_a-zA-Z0-9]*/  // => `Regex<Never>`
-print(str.firstMatch(of: identifier)?.captures)
-// ❗️ Program aborts or hangs.
+extension String {
+    public func firstMatch<R: RegexProtocol, C...>(of regex: R)
+        -> (Substring, C...)? where R.Captures == (C...)
+}
+
+// Expands to:
+//     extension String {
+//         func firstMatch<R: RegexProtocol>(of regex: R)
+//             -> Substring? where R.Captures == ()
+//         func firstMatch<R: RegexProtocol, C1>(of regex: R)
+//             -> (Substring, C1)? where R.Captures == (C1)
+//         func firstMatch<R: RegexProtocol, C1, C2>(of regex: R)
+//             -> (Substring, C1, C2)? where R.Captures == (C1, C2)
+//         ...
+//     }
 ```
 
-In contrast, using `Void` as the empty capture type would allow `captures` to be
-accessed safely at anytime. When a regex has no captures, the match result's
-capture is simply `()`.
+For simple regular expressions this had the benefit of aligning the generic
+signature more obviously with the captures in the regex.
 
 ```swift
-let identifier = /[_a-zA-Z]+[_a-zA-Z0-9]*/  // => `Regex<Void>`
-print(str.firstMatch(of: identifier)?.captures)
-// Prints `()`.
+let regex = /ab(cd*)(ef)gh/
+// => `Regex<(Substring, Substring)>`
 ```
 
-`()` is also just a more consistent, continuous representation of a type with 0
-captures:
+However, it came with a number of (not necessarily insurmountable) open
+questions:
 
-| Number | Capture type             |
-|--------|--------------------------|
-| 0      | `()`                     |
-| 1      | `(Substring)`            |
-| 2      | `(Substring, Substring)` |
+- Will variadic generic tuple splatting preserve element labels?
+- Will variadic generic tuple splatting eliminate `Void`s? _(We don't want
+`firstMatch(of:)` to return `(Substring, Void)` for a regex with no captures)._
+- Will we be able to add [single-element labeled tuples](https://forums.swift.org/t/single-element-labeled-tuples/9797)?
+_(This would be needed to preserve the name of a capture in a regex with a
+single named capturing group.)_
+- What should be the type of `Captures` for a regex with no captures (e.g.
+`Void` or `Never` or something else)?
 
-### `Regex<(EntireMatch, Capture...)>` instead of `Regex<Captures>`
+Also, when using a regex within the result builder syntax, it's more obvious
+that `capture()` will have access to the entire match, when the match is
+included in the generic signature.
```

Comment:
> it's more obvious that `capture()` will have access to the entire match

This point is not so obvious to me. In either the old or the new design, a regex with the result builder syntax can be captured to produce the its entire match in the pattern context.

```swift
let pattern = Pattern {
    /abc/.capture() // either design can capture the entire match of this regex.
}
```

Is this what you are referring to, or something else?

Reply:
I was thinking about a situation where the regex includes captures:

```swift
let pattern = Pattern {
   /([0-9A-Z]+)(?:\.\.([0-9A-Z]))?/.capture { (match, lower, upper) in ... }
}
```

If we think this point is valuable, then we probably need to include an example to clarify.

Reply:
Hmmm in this case it still follows that `capture()` will include the entire match as a `Substring` in both designs. That is, if you capture anything it gets included in the generic type in both approaches.